### PR TITLE
Assert cached decode matches recomputing without a cache

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48289) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48289)

Claude flagged this while i was doing some work on exporters.

`GenerationTesterMixin` already ran greedy generate with and without a cache, but each test only checked its own shapes — nothing compared them, so a cache that feeds its layers the wrong positions or a mask of the wrong width passed both. Driving each side through `generate` keeps the modality, encoder-decoder and input-naming plumbing in the library, which is what lets one assertion cover text, VLM and encoder-decoder models alike.

It compares per-step logits, then ids up to the first near-tie — tie-checking the scores rather than the logits, since the ids are picked after the processors run and `min_new_tokens` drives EOS to -inf. Models whose state *is* their cache skip on `_is_stateful`.

Across every model it found four real bugs:

- fsmt embedded positions from the sliced one-token input, so every decode step re-read position `padding_idx + 1`. On real weights the drift flips a token by step 6; an earlier commit had overwritten the expected en-ru translation to absorb it, and the fix restores the original text for all four pairs.
- axk2, deepseek_v32 and glm_moe_dsa fold their sparse top-k into the attention mask with `masked_fill`, which writes True — i.e. *attend* — when the mask is boolean, as it is on the sdpa path. Padded sequences longer than `index_topk` attended future tokens.
- paligemma (and v1) let generated tokens inherit the prompt's `token_type_ids`, putting them in the bidirectionally-attending prefix block, so without a cache earlier tokens attended tokens generated after them.
- vibevoice_asr dropped its audio inputs after prefill even with no cache, so the placeholders fell back to text embeddings.

blt (decode rebuilds hash n-grams and patch lengths from one byte and gives the global trunk no cache), vibevoice_asr's residual VAE sampling and prophetnet's collapsed single-step branch are skipped with the measured mechanism.

# What does this PR do?

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. These often are low-quality, or fix extremely minor issues that occur rarely or never in practice.
As a result, we're instituting a rule that **first-time contributors should not use code agents to submit PRs or issues**.
We'd also ask autonomous "OpenClaw"-like agents not to open any PRs or issues.

Issues/PRs from first-time contributors that violate this rule will probably just be closed without review, and we
might block you, especially if you open more than one or appear to be deliberately ignoring this. We especially do not
want new contributors to jump in on random issues to contribute an agent-written fix. This creates lots of noise
for reviewers and other users and will almost certainly get you blocked.

For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [ ] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -408,6 +407,60 @@ def test_greedy_generate_dict_outputs_use_cache(self):
...
+            # allow. Tie-check the *scores*, not the logits — the ids are picked from the scores, and the
+            # processors in between move them (`_greedy_generate` sets `min_new_tokens`, so EOS is -inf and
+            # the real contest can be between two candidates the raw logits rank third and fourth).
+            prompt_length = cached.sequences.shape[1] - len(cached.scores)
+            for step, scores in enumerate(cached.scores):
+                top_two = scores.topk(2, dim=-1).values
+                if (top_two[:, 0] - top_two[:, 1]).min() <= 2e-3:
+                    break
+                position = prompt_length + step
+                self.assertListEqual(cached.sequences[:, position].tolist(), cacheless.sequences[:, position].tolist())
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py]: maybe `assert_similar_generate_outputs` which check if two generations are similar in logits or have same output IDs, and accounts for possible noise?

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 2]
```diff
@@ -381,6 +382,29 @@ def forward(
...
+            # A generated token is the suffix, which attends causally. The generic update repeats the last
+            # prompt value instead — 0 for a prompt with no suffix yet — putting the new tokens in the
+            # bidirectional *prefix* block, so with no cache, where the mask is rebuilt over the whole
+            # grown sequence every step, earlier tokens end up attending tokens generated after them.
+            # `token_type_ids == 0` is also what the processor masks out of `labels`, so a predicted token
+            # is 1 by definition.
+            model_kwargs["token_type_ids"] = torch.cat(
+                [token_type_ids, token_type_ids.new_ones((token_type_ids.shape[0], num_new_tokens))], dim=-1
+            )
+        return model_kwargs
```
Comment (zucchini-nlp) [inline src/transformers/models/paligemma/modeling_paligemma.py:402]: ah great catch! nit, can we reword the comment, it is a bit too verbose and claude-style

File: src/transformers/models/git/modeling_git.py [hunk 3]
```diff
@@ -841,11 +836,11 @@ def forward(
                 attention_mask = torch.cat(
                     [torch.ones_like(image_token_type_ids, dtype=attention_mask.dtype), attention_mask], dim=-1
                 )
-        elif past_key_values is not None and seq_len == 1:
-            # Expand attention mask and cache position with image tokens because GIT doesn't add image
-            # placeholder tokens when processing. Doesn't worth the refactor, low usage!
+        elif past_key_values is not None and past_key_values_length > 0 and attention_mask is not None:
+            # GIT keeps the image tokens in the cache without placeholder tokens in `input_ids`, so the
+            # incoming padding mask is narrower than the cache — widen it over the cached image tokens.
```
Comment (zucchini-nlp) [inline src/transformers/models/git/modeling_git.py]: i suppose we have to also check if mask is not 4d, so `past_key_values_length > 0 and mask is not None and mask.ndim == 2`

File: src/transformers/models/git/modeling_git.py [hunk 4]
```diff
@@ -788,11 +788,6 @@ def forward(
                 else past_key_values.get_seq_length()
             )
 
-        # Adjust position ids by adding image seq length
-        seq_len = input_ids.shape[1] if input_ids is not None else inputs_embeds.shape[1]
-        if pixel_values is None and past_key_values is not None and seq_len == 1:
-            position_ids = position_ids + past_key_values_length
-
```
Comment (zucchini-nlp) [inline src/transformers/models/git/modeling_git.py:795]: positions are tracked correctly now?

File: src/transformers/models/gemma3/modular_gemma3.py [hunk 5]
```diff
@@ -767,6 +759,22 @@ class Gemma3ForConditionalGeneration(PaliGemmaForConditionalGeneration):
...
+        model_kwargs: dict[str, Any],
+        is_encoder_decoder: bool = False,
+        num_new_tokens: int = 1,
+    ) -> dict[str, Any]:
+        # Gemma3's `token_type_ids` are the multimodal ones, marking *image* spans (`== 1`, see
+        # `get_block_sequence_ids_for_mask`) rather than PaliGemma's prefix, so the generic update — which
+        # repeats the last value, 0 for the text a prompt ends on — is already right here. Spelled out
+        # against `GenerationMixin` in the modular source, where a plain `super()` delegation would instead
+        # be read as "inherit PaliGemma's body".
+        return GenerationMixin._update_model_kwargs_for_generation(
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma3/modular_gemma3.py]: same here, can we make less verbose comments

**Review by zucchini-nlp — COMMENTED** [review]
Quickly skimmed through multimodal ones, great catch on gemmas!

File: src/transformers/models/axk2/modeling_axk2.py [hunk 6]
```diff
@@ -624,6 +624,13 @@ def forward(
         if past_key_values is not None:
             key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx)
 
+        # Both the indexer's scores and the sparse fold below *add* into this mask, so it has to be
+        # additive. `create_causal_mask` hands back a boolean mask on the sdpa path, where `masked_fill`
+        # writes `True` — i.e. *attend* — into every key the indexer's top-k dropped.
+        if attention_mask is not None and attention_mask.dtype == torch.bool:
+            attention_mask = torch.zeros_like(attention_mask, dtype=hidden_states.dtype).masked_fill_(
+                ~attention_mask, torch.finfo(hidden_states.dtype).min
+            )
```
Comment (vasqu) [inline src/transformers/models/axk2/modeling_axk2.py]: Checking tomorrow in details but imo we should potentially just do the conversion once where we create the mask itself + pass the kwarg to the mask creation that disallows skipping the mask creation

1. Is just to avoid reconverting at each layer
2. Is to ensure we always receive the mask as causality in sdpa can be skipped within the is_causal flag but for the indexer we dont have that shortcircuit

Based on that, we probably can always assume a mask in the end. Tbh this might be big enough that I'd like a dedicated PR with its own test potentially

Comment (vasqu) [reply src/transformers/models/axk2/modeling_axk2.py]: Honestly we might even change up SDPA to always return float (then eager == sdpa as well) cc @Cyrilvallez if you have insights why SDPA kept a bool mask 👀 

But yea I'd love a separate PR for this because it should fix left padding for these models (we need to set the index topk lower in those testers)

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 7]
```diff
@@ -1108,19 +1107,14 @@ def make_positions(tensor, padding_idx: int):
         mask = tensor.ne(padding_idx).int()
         return (torch.cumsum(mask, dim=1).type_as(mask) * mask).long() + padding_idx
 
-    def forward(
-        self,
-        input,
-        incremental_state: Any | None = None,
-        timestep: Tensor | None = None,
-    ):
+    def forward(self, input, past_key_values_length: int = 0):
```
Comment (vasqu) [inline src/transformers/models/fsmt/modeling_fsmt.py]: Is this not breaking as we change signature?

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 8]
```diff
@@ -575,11 +574,11 @@ def forward(
...
             raise ValueError("You cannot specify both decoder_input_ids and decoder_inputs_embeds at the same time")
         elif input_ids is not None:
-            # embed positions
-            positions = self.embed_positions(input_ids)
-            if use_cache:
-                input_ids = input_ids[:, -1:]
-                positions = positions[:, -1:]  # happens after we embed them
+            # Embed positions. `generate` hands one token per step, so the tokens already in the cache have
+            # to be counted in — positions come from a cumsum over `input_ids`, which on a length-1 step
+            # restarts at `padding_idx + 1` and would give every decode step the same position.
```
Comment (vasqu) [inline src/transformers/models/fsmt/modeling_fsmt.py]: Too bloated comment imo, a simple `embed positions accounting for potential past seq len` suffices no?

File: src/transformers/models/gemma3/modular_gemma3.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/models/gemma3/modular_gemma3.py]: I'd go further and do 
```
# Paligemma's interpretation for `token_type_ids` is unique and doesn't hold for our usual interpretation that appends texts step by step afterwards
def _update_model_kwargs_for_generation(self, **super_kwargs): 
    raise AttributeError()  # <- this avoids inheriting as per modular
```
I guess the end goal is to have the normal generation mixin's super than the paligemma one? I just don't see why we need the explicit super because in modeling we have the normal "generation mixin" --> so just let the normal super be enacted no?

Comment (vasqu) [reply src/transformers/models/gemma3/modular_gemma3.py]: After digging again, paligemma doesnt even override this so the comment even makes less sense to me; maybe im missing something?

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/paligemma/modeling_paligemma.py:402]: Ah ok so we update paligemma sorry now it makes sense but then my comment about not inheriting would be better imo

File: src/transformers/models/vibevoice_asr/modular_vibevoice_asr.py [hunk 9]
```diff
@@ -462,7 +462,10 @@ def prepare_inputs_for_generation(self, *args, is_first_iteration=False, **kwarg
 
         model_inputs = super().prepare_inputs_for_generation(*args, **kwargs)
 
-        if is_first_iteration:
+        # Without a cache, every step recomputes the whole prefix, so the audio has to come along or the
+        # placeholder tokens silently fall back to their text embedding — the same rule the base
+        # implementation applies through `MULTIMODAL_INPUTS_TO_DROP_OUTSIDE_PREFILL`.
+        if is_first_iteration or not kwargs.get("use_cache", True):
```
Comment (vasqu) [inline src/transformers/models/vibevoice_asr/modular_vibevoice_asr.py:465]: I dont even think we need the comment at all

File: tests/generation/test_utils.py (hunk 1 above)
Comment (vasqu) [reply tests/generation/test_utils.py]: Agree I think generation is a good indicator itself with the mentioned fn. Not sure whether logits could bite us by being flaky

File: tests/models/git/test_modeling_git.py [hunk 10]
```diff
@@ -530,7 +530,7 @@ def test_inference_image_captioning(self):
         self.assertEqual(outputs.sequences.shape, expected_shape)
         self.assertEqual(generated_caption, "two cats laying on a pink blanket")
         self.assertTrue(outputs.scores[-1].shape, expected_shape)
-        expected_slice = torch.tensor([-0.8126, -0.8123, -0.8119], device=torch_device)
+        expected_slice = torch.tensor([-0.8800, -0.8798, -0.8794], device=torch_device)
```
Comment (vasqu) [inline tests/models/git/test_modeling_git.py:533]: Is this not an indication that the behavior is changed or was it generating without cache here?

**Review by vasqu — COMMENTED** [review]
Proper review this time. let's use the modular trick to not inhert instead of using the GenerationMixin.super(...)

Other than that, I'd love to split the indexer to a different PR because it is even bigger than this imo

File: tests/generation/test_utils.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply tests/generation/test_utils.py]: assert_similar_generate_outputs seems to be very loose, similar ids don't mean the computations are similar, especially when one generation only gets a diff mask. for example the axk2 fails the logits check without the indexer fix @vasqu suggested (0.0187 > 1e-2) but survives assert_similar_generate_outputs

Comment (vasqu) [reply tests/generation/test_utils.py]: Oof ok, so we do need the logits :cry: let's keep a logits comparison then but keep an eye out for flakiness

File: tests/models/git/test_modeling_git.py (hunk 10 above)
Comment (IlyasMoutawwakil) [reply tests/models/git/test_modeling_git.py:533]: yep the old output was the bugged one

Comment (IlyasMoutawwakil) [reply tests/models/git/test_modeling_git.py:533]: and the generation in this test didn't change, so it was also a sub-argmax-changing bug 😭

File: src/transformers/models/fsmt/modeling_fsmt.py (hunk 7 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/fsmt/modeling_fsmt.py]: yes but the removed arguments were never read actually

Comment (vasqu) [reply src/transformers/models/fsmt/modeling_fsmt.py]: Can still make BC friendly tho with e.g. "*" no?

**Review by vasqu — APPROVED** [review]
Looking good, let's wait for #48363 first then merge this

Also @zucchini-nlp if you also want to have one last sanity check

**Comment by vasqu** [comment]
Hmm, dont wanna invest much time into this but it's mergable when you come back @IlyasMoutawwakil just need to fix / skip the last model(s)

**Comment by IlyasMoutawwakil** [comment]
run-slow: blt, fsmt, gemma3, gemma3n, git, paligemma, prophetnet, vibevoice, vibevoice_asr

**Comment by IlyasMoutawwakil** [comment]
@vasqu seems to work for all models

**Comment by IlyasMoutawwakil** [comment]
will merge once green and slow tests are unaffected

**Comment by IlyasMoutawwakil** [comment]
spoke too fast, Kosmos2_5ModelTest.test_cached_decode_matches_cacheless is failing

File: tests/models/fsmt/test_modeling_fsmt.py [hunk 11]
```diff
@@ -495,10 +495,6 @@ def translation_setup(self, pair):
 
         src_text = text[src]
         tgt_text = text[tgt]
-        # To make `test_translation_pipeline_0_en_ru` pass in #38904. When translating it back to `en`, we get
-        # `Machine learning is fine, isn't it?`.
-        if (src, tgt) == ("en", "ru"):
-            tgt_text = "Машинное обучение - это прекрасно, не так ли?"
```
Comment (IlyasMoutawwakil) [inline tests/models/fsmt/test_modeling_fsmt.py:501]: this was a workaround that's no longer needed

**Comment by IlyasMoutawwakil** [comment]
fsmt test is fixed now
Kosmos2_5ModelTest.test_cached_decode_matches_cacheless passes locally, might be flaky

**Comment by IlyasMoutawwakil** [comment]
run-slow: fsmt
